src/freertos_drivers/common/WiFiInterface.hxx

```diff
@@ -0,0 +1,250 @@
+    {
+        /// Access point has no security.
+        OPEN,
+        /// Access point uses WEP security.
+        WEP,
+        /// Access point uses WPA2 security.
+        WPA2,
+    };
+
+    /// Status of the WiFi interface.
+    enum class WiFiState : uint8_t
+    {
+        /// The interface has not been initialized or is otherwise not in a
+        /// usable state. A call to @ref start would be required. This is the
+        /// default state of the interface when the WiFiInterface instance is
+        /// created.
+        OFF = 0,
+        /// The interface is in an IDLE state (not STATION or ACCESS POINT) but
+        /// has been initialized. This is the expected state after calling
+        /// @ref stop. A call to @ref start would be required to prepare the
+        /// interface for use.
+        IDLE,
+        /// The interface is currently in an initializing state. It is not
+        /// expected to be usable at this point.
+        INITIALIZING,
+        /// The interface is currently connecting to an access point.
+        CONNECTING,
+        /// Interface is not associated with an access point. This state can be
+        /// used in the event the interface has lost it's connection to the
+        /// access point.
+        NOT_ASSOCIATED,
+        /// Interface has not yet received an IP address from the connected
+        /// access point.
+        NO_IP,
+        /// The interface is in normal operating status and is ACTIVE as either
+        /// a STATION, ACCESS POINT or both (if supported).
+        ACTIVE
+    };
+
+    /// WiFi Interface Operating mode.
+    enum class WiFiMode : uint8_t
+    {
+        /// Interface mode is not known.
+        UNKNOWN = 0,
+        /// Interface operates as a STATION (connects to access point).
+        STATION,
+        /// Interface operates as an ACCESS POINT (other STATIONS can connect
+        /// to it).
+        ACCESS_POINT,
+        /// Interface will operate as both a STATION and ACCESS POINT. If the
+        /// interface does not support this mode an error will be raised by the
+        /// @ref start method.
+        STATION_AND_ACCESS_POINT
+    };
+
+    /// Contains the metadata for an access point.
+    struct NetworkListEntry
+    {
+        /// SSID of the access point.
+        std::string ssid;
+        /// Security type of the access point.
+        WiFiSecurity security;
+        /// Receive Signal Strength Indicator of the access point.
+        int rssi;
+    };
+
+    /// Startup the WiFi interface.
+    ///
+    /// @param mode is the requested mode of the WiFi interface.
+    /// @return zero upon successfully starting in the provided mode, non-zero
+    /// otherwise.
+    int start(WiFiMode mode = WiFiMode::STATION);
+
+    /// Stops the WiFi interface in preparation for a reboot.
+    void stop();
+
+    /// @return true if WiFi interface has been started, false otherwise.
+    bool is_started();
+
+    /// Connect to access point.
```

Comment:
adding a timeout parameter to make this clear that it is a synchronous (blocking) call. Though, it may be up to the implementation weather or not it is truly blocking. For code simplicity though blocking makes more sense for this API.